### PR TITLE
Support the HTML5 required attribute in MSHTML documents

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
As far as I know, this pr has no dedicated issue for itself. This pr is based on findings in https://github.com/nvaccess/nvda/issues/7320#issuecomment-310827282

### Summary of the issue:
NVDA does not report the required state of a html element when the HTML5 required attribute is used for this.

### Description of how this pull request fixes the issue:
Added support for required in de NVDAHelper MSHTML vbuf backend and associated python code.

### Testing performed:
For required

```
<input id="name" required type="text" />
```

For not required

```
<input id="name" type="text" />
```

Note that required="false" will still result in required state being added, but this is no bug according to [the spec](https://www.w3schools.com/tags/att_input_required.asp). I've seen the same behaviour in Firefox.

### Change log entry:
I consider the absence of this feature to be a bug.

* Bug fixes
    + In MSHTML documents, the HTML5 required attribute is now supported to indicate the required state of a form field (#7321)

### CODE REVIEW

**Comment by michaelDCurran** [comment]
This looks good so far, but can you please also support this in focus mode. See aria-required in ```NVDAObjects\IAccessible\MSHTML.py```

**Comment by leonardder** [comment]
Hmm, good point. However, it seems this is going to be more difficult than I thought

```
<input type="text" required="required" name="cheese" />
```

```
>>> nav.HTMLNode.getAttribute('type')
u'text'
>>> nav.HTMLNode.getAttribute('name')
u'cheese'
>>> type(nav.HTMLNode.getAttribute('required'))
<type 'NoneType'>
```

It seems the node doesn't expose the required attribute correctly.

**Comment by leonardder** [comment]
A, it turns out that I can use HTMLNode.hasAttribute('required')

**Comment by leonardder** [comment]
@michaelDCurran: this is now working.

I added a __contains__ method to the HTMLAttribCache to support stuff like 'required' in self.HTMLAttributes. I decided to give it a separate cache, as I didn't want to change the behaviour of __getitem__. This nieuw _contains__ method fixes a bug where 'attribute' in HTMLAttributes resulted in a freeze of NVDA.

File: source/NVDAObjects/IAccessible/MSHTML.py [hunk 1]
```diff
@@ -71,6 +71,7 @@ class HTMLAttribCache(object):
 	def __init__(self,HTMLNode):
 		self.HTMLNode=HTMLNode
 		self.cache={}
+		self.containsCache={}
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/MSHTML.py]: This can be a set rather than a dictionary

File: source/NVDAObjects/IAccessible/MSHTML.py [hunk 2]
```diff
@@ -84,6 +85,23 @@ def __getitem__(self,item):
 		self.cache[item]=value
 		return value
 
+	def __contains__(self,item):
+		try:
+			return self.containsCache[item]
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/MSHTML.py]: Then change this to  return item in self.containsCache

File: source/NVDAObjects/IAccessible/MSHTML.py [hunk 3]
```diff
@@ -84,6 +85,23 @@ def __getitem__(self,item):
 		self.cache[item]=value
 		return value
 
+	def __contains__(self,item):
+		try:
+			return self.containsCache[item]
+		except LookupError:
+			pass
+		value = self[item]
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/MSHTML.py]: I would perhaps  only check item in self.containsCache, then item in self.cache, and if none of them then node.hasattribute(item) and add the answer to containsCache.  hasattribute should work for any value that getattribute would work for.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Comment by michaelDCurran** [comment]
Make sure to test on several versions of Internet Explorer. At very least what ever Internet Explorer comes with Windows 7.

File: source/NVDAObjects/IAccessible/MSHTML.py (hunk 1 above)
Comment (leonardder) [inline source/NVDAObjects/IAccessible/MSHTML.py]: Mm, not sure whether we're on the same track here. Changing this to a set means that we lose track of the attributes for which hasAttribute has returned False in the past, so in other words, False evaluations won't get cached. We either need a dictionary for this, or have two sets for hasAttribute True and hasAttribute False. Could it be that you overlooked this, or is it your intention not to cache the Falses?

File: source/NVDAObjects/IAccessible/MSHTML.py (hunk 2 above)
Comment (leonardder) [inline source/NVDAObjects/IAccessible/MSHTML.py]: If we do this, the block below that actually runs hasAttribute never gets executed.

**Comment by leonardder** [comment]
This is a good point. I hope I have a proper vm around for this.

**Comment by leonardder** [comment]
@michaelDCurran: Would you be able to explain why you belief containsCache should be a set rather than a dictionary? See https://github.com/nvaccess/nvda/pull/7321#discussion_r125560509

If you agree with containsCache being a dictionary anyway, this is ready for another review.

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
This looks good to me. Ignore the set/dictionary comment.

**Comment by michaelDCurran** [comment]
@leonardder  If you are finnished with this I'll incubate it?

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [19 jul. 2017 03:37 CEST](https://github.com/nvaccess/nvda/pull/7321#issuecomment-316246838 "2017-07-19T01:37:04Z - Replied by Github Reply Comments"):
> @leonardder  If you are finnished with this I'll incubate it?

Yes, go ahead. Thanks a lot!

**Pull request merged**